src/interfaces/sampler_interface.jl

```diff
@@ -0,0 +1,125 @@
+#################
+# The sampler interface
+#
+# In this file, we define the interface for general samplers and specific
+# samplers for scattering processes.
+#################
+
+"""
+Abstract base type for general sampler.
+
+!!! note "Sampler interface"
+
+    Functions to be implemented
+
+    ```Julia
+    CustomSampler <: AbstractSampler{Float64}
+
+    Base.eltype  # type of the generated samples
+    QEDevents._weight
+    Base.size  # Return the dimensionality of the sampler (number of outgoing four-momenta per sample).
+    is_exact
+    QEDevents._rand!(rng::AbstractRNG, ::AbstractSampler, x::AbstractVector{T}) where {T}
+
+    ```
+    
+    optional
+
+    ```Julia
+    _post_processing
+    ```
+
+"""
+abstract type AbstractSampler <: AbstractComputationSetup end
+
+@inline Base.eltype(::AbstractSampler) = Float64
+
+"""
+Interface function, which asserts that the given `input` is valid.
+"""
+function _assert_valid_input(smplr::AbstractSampler, x::AbstractVecOrMat)
+    size(x, 1) == size(smplr, 1)[1] || error(
+        "Invalid input. The dimensionality of the input is $(size(x,1)) but it should be $(size(smplr,1)[1]).",
+    )
+    return eltype(x) == eltype(smplr) || error(
+        "Invalid input. The element type of the input is $(eltype(x)) but it should be $(eltype(smplr)).",
+    )
+end
+
+function _compute(smplr::AbstractSampler, sample)
+    return _weight(smplr::AbstractSampler, sample)
+end
+
+"""
+
+    _weight(::AbstractSampler, sample)
+
+Return the weight associated with the given sample according to the sampler.
+"""
+function _weight end
+
+"""
+
+    setup(::AbstractSampler)
+
+???
+"""
+function setup end
+
+"""
+
+    is_exact(::AbstractSampler)
+
+Return if the sampler is exactly representing the base distribution or not.
```

Comment:
The docstring should indicate, that this is an interface function.

Reply:
Added in e91af7c76038a6cbcaeb764333ea07da0e5bacab